### ISSUE TITLE
Factor out `.pest` parser for use by other tools

### ISSUE THREAD

**Issue by Victor-Savu** [pest-parser/pest#158]
I was wondering if it would make sense to factor out the part of the code which (bootstrappingly?) parses the `.pest` file so that other tools such as syntax highlighters and completion engines could use it.

**Comment by dragostis** [comment]
I very much agree. When I first started working on this, I was not aware that `proc_macro` crates had some limitations. In order to get highlighting and completion, one can use `intellij-rust`, but clippy seems to completely ignore this, so it makes sense to do this.

Would you be willing to give this a try?

**Comment by dragostis** [comment]
Actually, factoring this out in a separate crate would also mean making a lot of APIs which are not yet documented public, which means that there will be a bit more work to do.

**Comment by Victor-Savu** [comment]
Ohh right that kind of completion  :) I was referring to to highlighting and completion of the grammar specification language used in the `.pest` file.

**Comment by Victor-Savu** [comment]
This should be a really nice self-contained project, and I'd be happy to have a go at it.

**Comment by dragostis** [comment]
Oh, oh! That too, of course. Hm. It would be great if there was a way to bootstrap pest in order to have the grammar written in pest form, but I don't think that's actually possible.

Yes, I think it can work as a separate project and I'm willing to mentor and help out. :)

**Comment by Victor-Savu** [comment]
Yep, the bootstrapping is that I am working on now :) The past 3 hours have shown that it is not exactly straightforward, but bootstrapping never is. I'll just try and see what comes out.

**Comment by Victor-Savu** [comment]
[WIP comment - I just don't want the art work below to be accidentally lost]

This is great! I would really like to learn the code, philosophy and in general what your vision is regarding the roles of the different modules before committing to a particular architecture. I hope you don't mind if I ask some silly questions along the way.

From your suggestion, I understand that `pest_derive::{parser, ast, validator, optimizer, generate}` all deal with "compiling" the `pest` language:
```
      [FileInput]
           |
           V
(parser + ast + validator) --> [panic!]
           |
    [Vec<ast::Rule>]
           |
           V
      (optimizer)
           |
           V
    [Vec<ast::Rule>]
           |
           V
      (generator)
           |
    [quote::Tokens]
```
If my ASCII doodle is accurate, my initial thoughts would be:
- The `parser + ast + validator` bundle hold the most value for external use. One immediate usability improvement would be if the `validator` returned the errors instead of panicking, leaving the external application to decide on the appropriate reaction to the error.
- I see the `optimizer` as being valuable (perhaps) for an outside linter, which would suggest static improvements. In this state, however, the external application would have a hard time diff-ing the input and the output of the `optimizer` to figure out what changed. One solution could be if the optimizer was split into two segments:
  - The former takes the set of rules (by reference) and outputs a list of suggested edits
  - The latter takes the set of rules and the set of edits and outputs what the current `optimizer` does
  The former segment can be used by third party linters to suggest optimizations.
- The `generator` be really specific to generating Rust tokens. Personally, I don't see any uses for it outside of the `pest_derive` crate.

I would propose a two-step approach:
- Factor out the `parser + ast + validator` into a separate crate, the name of which would be the subject of most intense bikesheding

**Comment by Victor-Savu** [comment]
Answering to the first three parts of [your commend above]:
> - this makes sense to have its own crate/repo; probably a good name would be pest_meta
> - this should house the parser, validator, optimizer and generate that are currently in pest_derive
> - pest_derive should only make use of the API

I would propose a two-step approach:
1. Factor out the `parser + ast + validator` into a separate crate, the name of which would be the subject of most intense bikesheding, as:
   - `pest_meta` is too general and people would not immediately grasp its intended purpose
   - `pest_syntax` is confusing for a library that deals with parsers, which would make it __meta__, therefore justifying the name `pest_meta`

   Part of this refactoring would include a change in the interface such that the syntax errors are returned rather than `panic!`ed out.

1. Factor out the `optimizer`, either into the same crate as before or into a crate called `pest_how_to_optimally_name_stuff` if the [split proposed above](https://github.com/pest-parser/pest/issues/158#issuecomment-331732787)is feasible. IMHO, this can happen a bit later, since the value proposition is somewhat smaller than that of the first part.

1. [index-out-of-bounds] Leave the `generator` in `pest_derive` until someone has another use for it.

From the point of view of `pest_derive` crate users, there would be absolutely no change, so this is all fully backwards compatible and at least the first step is quite unlikely to introduce any regressions.

**Comment by Victor-Savu** [comment]
Answering to the last two parts of [your commend above](https://github.com/pest-parser/pest/issues/158#issuecomment-331722332):
> - it would be nice to experiment with better bootstrapping, i.e. translation of the manual pest parser to a generated one
With the parser in a separate crate, bootstrappig should at least make for a fun project :smile:
> - it would probably make sent to use the Language Server protocol, since providing completion should be extremely easy; this should probably be a binary within `pest_meta`
I am totally :+1: for the [Language Server Protocol](https://github.com/Microsoft/language-server-protocol). I would, however, leave it to another crate, as that might require some dependencies that other users of the crate (e.g. `pest_derive`) might not want to take on.

**Comment by Victor-Savu** [comment]
I'm really excited to hear your thoughts!

Also, I have been meaning to say this for a while, and now having looked at the code: __Awesome work with Pest & Pest Derive!__ It's such a great example of how to make a crate user- and contributor- friendly!

**Comment by dragostis** [comment]
Sorry for the late reply, but here are my thoughts:

> * The parser + ast + validator bundle hold the most value for external use. One immediate usability improvement would be if the validator returned the errors instead of panicking, leaving the external application to decide on the appropriate reaction to the error.

I see the value in this. I think a good approach here would be to decouple the parser from the validator a bit more, since right now, they're quite close. The sudden panicking is not by design; it's a shortage that came while winning some pragmatism, and refactoring so as to remove the panics is a good direction.

> * I see the optimizer as being valuable (perhaps) for an outside linter, which would suggest static improvements. In this state, however, the external application would have a hard time diff-ing the input and the output of the optimizer to figure out what changed. One solution could be if the optimizer was split into two segments:
>   * The former takes the set of rules (by reference) and outputs a list of suggested edits
>   * The latter takes the set of rules and the set of edits and outputs what the current optimizer does
>   * The former segment can be used by third party linters to suggest optimizations.

Hm, I'm not so sure about this. The current optimizer work without asking you to change the grammar in any way, and I'm not quite sure whether this might be the case in the future or not. The optimizer and generator may as well continue their life in the derive crate.

What would probably me of more value would be for the parser to have *some* degree of flexibility when it comes to parsing. Partial parsing might come in handy and would be an interesting area to research with `pest`.

> `pest_meta` is too general and people would not immediately grasp its intended purpose

I'm open to better suggestions, but I don't think that the people actually willing to use `pest_meta` would find it confusing. The reason why I like meta is because it's ambiguous enough to handle a lot of stuff that we might be interested in adding later on. (like generation, maybe?)

> I am totally :+1: for the [Language Server Protocol](https://github.com/Microsoft/language-server-protocol). I would, however, leave it to another crate, as that might require dependencies that other users of the crate (e.g. `pest_derive`) might not want to take on.

This sounds like a fair approach, yes.

> Also, I have been meaning to say this for a while, and now having looked at the code: Awesome work with Pest & Pest Derive! It's such a great example of how to make a crate user- and contributor- friendly!

Glad to hear that. I feel like the approach could be handled a bit better, but I'm giving my best to get there post-`1.0`.

To sum up, a good starting point would be to have a separate crate with some bootrapping working. If not, the current parser in it's raw state should also do fine.

**Comment by Victor-Savu** [comment]
The proposal regarding the optimizer was a bit far-reaching anyhow :smile: But just to clarify regarding your remark:
> The current optimizer work without asking you to change the grammar in any way, and I'm not quite sure whether this might be the case in the future or not.

The idea was that the optimizer applies edits from a similar perspective that a linter would have when proposing code improvements: The code works as-is, but it would be _"better"_ with the given suggestions applied. Now, of course, an application interested in optimizing the code (without human interaction) would be fine with just applying the changes without informing the user of the applied changes. However, I imagined that some of the same knowledge used by the optimizer could be beneficial to someone editing the code, as usually these normalizations reveal simplified ways of thinking about the code. The mental model I have for this is [clang-tidy](http://clang.llvm.org/extra/clang-tidy/). That being said, it is far-fetched and would need a lot more experience before being put to good use, so I'm all for:
> The optimizer and generator may as well continue their life in the derive crate.

Regarding the name, I like `pest_meta` as well, but I don't necessarily agree (yet) with the possibility of adding more stuff. This might be personal preference (hard to tell sometimes), but I can't shake the feeling that users should be able to choose what they need: a parser, a generator, an optimizer. Taking the example of `pest_derive`, which, in essence, is a generator to Rust, it would not necessarily benefit from `pest_meta` providing other generators. Unless you are thinking about a more general interface for what a generator is, independent of the generation target, in which case, :+1: (though I can't really imagine what that would look like). Regardless, `post_meta` it is :)

**Comment by dragostis** [comment]
I guess the generator idea was not very fleshed out and probably not the best example. What I meant is that some extra functionality might find its way in the crate, but I don't yet know what that could be.

Would you like a repo on the `pest-parser` org? Or would it be better to transfer once done base is done from your own account?

**Comment by Victor-Savu** [comment]
I was thinking I would just work on my fork and then open a PR: I would put `pest_meta` right next to `pest_generate` and company and add it to the workspace. Would that work for you? I will try to rebase often so that the PR would reflect all of the upstream changes.

**Comment by dragostis** [comment]
`pest_drive` you mean? :P

It could work, but I feel like having it in a separate repo would also work.

**Comment by Victor-Savu** [comment]
yes, I meant `pest_derive` :) But all the development would happen in my fork and would only appear in upstream via PR.

**Comment by dragostis** [comment]
@Victor-Savu, have you managed to make any progress? Lemme know if you need a hand.

**Comment by Victor-Savu** [comment]
@dragostis sorry, having some problems in the non-digital world, but I will try to open a PR this weekend at least for the first part :)

**Comment by Victor-Savu** [comment]
I think this has been solved now :)

**Comment by dragostis** [comment]
@Victor-Savu I've made some interesting progress lately with the help of your patch. Hit me up on Gitter if you're curious to see.
